keanu-project/src/main/java/io/improbable/keanu/tensor/INDArrayShim.java

```diff
@@ -99,6 +175,250 @@ private static INDArray broadcastMinus(INDArray a, INDArray b) {
         }
     }
 
+    public static INDArray pow(INDArray left, INDArray right) {
```

Comment:
Great work here but there's a lot of duplicated code! I think we can write some functions that generalise this rank 1 stuff for us. We're doing the same steps each time:

1) calculate shape
2) operation if left is length 1
3) operation if right is length 1
4) return it

I've had a quick go at one:

```
    private static INDArray lengthOne(INDArray left, INDArray right, BiFunction<INDArray, INDArray, INDArray> func, double leftScalar, double rightScalar) {
        long[] resultShape = Shape.broadcastOutputShape(left.shape(), right.shape());
        INDArray result;
        if (left.length() == 1) {
            result = func.apply(Nd4j.valueArrayOf(right.shape(), leftScalar), right);
        } else {
            result = func.apply(left, Nd4j.valueArrayOf(left.shape(), rightScalar));
        }
        return result.reshape(resultShape);
    }
```

That we can call with:

```
    public static INDArray pow(INDArray left, INDArray right) {
        if (left.length() == 1 || right.length() == 1) {
            return lengthOne(left, right, INDArrayShim::pow, left.getDouble(0), right.getDouble(0));
        }
        return Transforms.pow(left, right, false);
    }
```

Have a play around and see if we can extend it to the other ones!

Reply:
Done it as far as I could.